third_party/amd/lib/TritonAMDGPUTransforms/InThreadTranspose.cpp

```diff
@@ -0,0 +1,620 @@
+  ForwardSearchAnalysis result;
+  for (auto &use : val.getUses()) {
+    auto user = use.getOwner();
+    LDBG("      processing user " << *user);
+    if (isa<triton::ReturnOp>(user)) {
+      LDBG("    Reached return from function");
+      return failure();
+    } else if (isa<scf::YieldOp>(user)) {
+      auto opIdx = use.getOperandNumber();
+      auto parent = user->getParentOp();
+      // traverse outside data flow
+      auto parentResult = parent->getResult(opIdx);
+      auto cfSearch = traverseCFForValueUses(parentResult);
+      if (failed(cfSearch))
+        return failure();
+      result.ops.append(cfSearch.value().ops);
+      result.transitiveCF.push_back(parentResult);
+      result.transitiveCF.append(cfSearch.value().transitiveCF);
+
+      // traverse loop internal data flow
+      if (auto forOp = dyn_cast<scf::ForOp>(parent)) {
+        int forBodyOperandIdx = opIdx + forOp.getNumInductionVars();
+        auto blockArg = forOp.getBody()->getArgument(forBodyOperandIdx);
+        auto cfSearch = traverseCFForValueUses(blockArg);
+        if (failed(cfSearch))
+          return failure();
+        result.ops.append(cfSearch.value().ops);
+        result.transitiveCF.push_back(blockArg);
+        result.transitiveCF.append(cfSearch.value().transitiveCF);
+      }
+    } else if (auto forOp = dyn_cast<scf::ForOp>(user)) {
+      LDBG("      for op num operands: " << forOp.getNumOperands());
+      LDBG("      for op body num operands: "
+           << forOp.getBody()->getNumArguments());
+      assert(use.getOperandNumber() >= forOp.getNumControlOperands());
+      int blockArgIdx = use.getOperandNumber() - forOp.getNumControlOperands() +
+                        forOp.getNumInductionVars();
+      auto blockArg = forOp.getBody()->getArgument(blockArgIdx);
+      auto cfSearch = traverseCFForValueUses(blockArg);
+      if (failed(cfSearch))
+        return failure();
+      result.ops.append(cfSearch.value().ops);
+      result.transitiveCF.push_back(blockArg);
+      result.transitiveCF.append(cfSearch.value().transitiveCF);
+    } else {
+      result.ops.push_back(user);
+    }
+  }
+  return result;
+}
+
+/// Look for defining operation, hopping over control flow.
+///
+/// Gather all operations of type T within one def-use hop from val,
+/// control flow constructions are not considered as an operations.
+/// \returns true on success, false if analysis failed
+template <typename Op>
+FailureOr<SmallVector<Op>> findAllDefiningOps(Value val) {
+  auto candidates = traverseCFForValueDefs(val);
+  if (failed(candidates))
+    return failure();
+  SmallVector<Op> result;
+  for (auto candidateValue : candidates.value()) {
+    auto op = candidateValue.getDefiningOp();
+    if (!op)
+      continue;
+    if (auto typedOp = dyn_cast<Op>(op))
+      result.push_back(typedOp);
+  }
+  return result;
+}
+
+/// Look for all operations with one of OpTy types in def-use chains in both
```

Comment:
What's `OpTy` here? Overall it looks this documentation is out of date given it doesn't match with the implementation.